### PR TITLE
[NativeAOT] allow creating CCW for resurrected objects

### PR DESCRIPTION
**Pull request by AustinWise**
This PR adds a test that on main passes for CoreCLR but segfaults for NativeAOT. It also includes fixes to make the test pass for NativeAOT.

The key problem is the conditional weak table will keep the `ManagedObjectWrapperHolder` alive but not prevent it from being finalized. This means that if the wrapped object is resurrected, calls to `GetOrCreateComInterfaceForObject` could return a pointer freed memory.

The solution this PR implements is to move the finialization of the `ManagedObjectWrapper` to a separate class. This class defers finialization until the wrapped object has been completely collected. The new `ManagedObjectWrapperReleaser` is referenced by the `ManagedObjectWrapperHolder` class, so it should not be finialized until the wrapped object becomes eligible for finialization.

The references between objects looks like this:

```mermaid
graph TD
WO[wrapped object];
MOW["ManagedObjectWrapper (unmanaged struct)"];
MOWH[ManagedObjectWrapperHolder];
Releaser[ManagedObjectWrapperReleaser];
WO -->|conditional weak table| MOWH;
MOWH --> WO;
MOWH --> MOW;
MOW -->|ref-counted handle| MOWH;
Releaser --> MOW;
MOWH --> Releaser;
```

Some other race conditions are also fixed by this PR:
* Previously the ref-counted handle in the `ManagedObjectWrapper` was not set until after the `ManagedObjectWrapperHolder` was added to the CWT. Hypothetically if a holder was added to the CWT but not yet initialized with the ref-counted handle, another thread could read the object out of the CWT and observe the null handle. This is fixed by doing all the initialization of the holder in its constructor.
* The `GetOrCreateComInterfaceForObject` was assuming that `ConditionalWeakTable.GetValue` always returns a new object and calling `AddRef` on the holder was not needed. However, `GetValue` can potentially call `TryGetValue` and return an existing instance. This means that if many threads called `GetOrCreateComInterfaceForObject` at the same time, the ref count on the `ManagedObjectWrapper` might not be as high as it should be. This is fixed by unconditionally calling `AddRef` on the holder.

### CODE REVIEW

**Comment by AustinWise** [comment]
FYI: this is still a draft because while this PR does make the tests pass, it introduces race conditions. Specifically there is no coordination between a thread resurrecting the holder & calling `AddRef` with the finalizer thread. So it is still possible for the strong reference to be released on the wrapped object while the COM reference count is non-zero.

**Comment by AustinWise** [comment]
Instead of having one object for managing lifetimes of both the managed and unmanaged objects, I split the responsibilities into two objects. I think this is closer to a solution, but I have to review it a bit more and write down my assumptions about how GCHandle work.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs [hunk 1]
```diff
@@ -404,42 +404,62 @@ static bool IsRootedCallback(IntPtr pObj)
...
             }
 
             public unsafe IntPtr ComIp => _wrapper->As(in ComWrappers.IID_IUnknown);
 
-            public object WrappedObject => _wrappedObject;
-
-            public uint AddRef() => _wrapper->AddRef();
+            public uint AddRef(object wrappedObject)
+            {
+                _exposed = true;
```
Comment (AustinWise) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:440]: I'm not sure if I need more synchronization on this variable. It is set to true on any random thread and read from the finializer thread.

File: src/tests/Interop/COM/ComWrappers/API/Program.cs [hunk 2]
```diff
@@ -336,6 +336,62 @@ unsafe static void CallSetValue(TestComWrappers wrappers, Test testInstance, int
...
+            {
+                Assert.NotEqual(null, Test.Resurrected);
+                IntPtr nativeInstance = wrappers.GetOrCreateComInterfaceForObject(Test.Resurrected, CreateComInterfaceFlags.None);
+                Assert.NotEqual(IntPtr.Zero, nativeInstance);
+
+                var iid = typeof(ITest).GUID;
+                IntPtr itestPtr;
+                Assert.Equal(0, Marshal.QueryInterface(nativeInstance, ref iid, out itestPtr));
+
+                var inst = Marshal.PtrToStructure<VtblPtr>(itestPtr);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/ComWrappers/API/Program.cs:381]: We shouldn't be using `Marshal.PtrToStructure()` anywhere in here. Please defer to pointers and remove the other two uses as well.

**Comment by AustinWise** [comment]
> Based on a quick look at the test I'm not sure I understand what this is validating.

On the main branch, the test passes on CoreClr but segfaults on NativeAOT. This is because the `ManagedObjectWrapperHolder` gets finalized and frees the `ManagedObjectWrapper` too early. I have updated the description of the PR to be a little more clear about this.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs [hunk 3]
```diff
@@ -404,42 +404,62 @@ static bool IsRootedCallback(IntPtr pObj)
...
+                GC.KeepAlive(this);
+                return ret;
+            }
 
             ~ManagedObjectWrapperHolder()
             {
+                if (_exposed && _wrappedObject.Target != null)
+                {
+                    // The wrapped object has not been fully collected, so it is still
+                    // potentially reachable via the CWT. Keep ourselves alive in case
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:451]: What is the CWT acronym?

Comment (jkoritzinsky) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:451]: ConditionalWeakTable

Comment (AaronRobinsonMSFT) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:451]: I see. There is no need to use an acronym here.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs (hunk 1 above)
Comment (AustinWise) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs]: I'm just going to remove this for now. I'll revisit the issue of extra memory being used when there is a race condition inserting into the conditional weak table later.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs (hunk 3 above)
Comment (AustinWise) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs]: I fixed the comment.

File: src/tests/Interop/COM/ComWrappers/API/Program.cs (hunk 2 above)
Comment (AustinWise) [reply src/tests/Interop/COM/ComWrappers/API/Program.cs]: I switch to using pointers.

File: src/tests/Interop/COM/ComWrappers/API/Program.cs [hunk 4]
```diff
@@ -336,6 +336,62 @@ unsafe static void CallSetValue(TestComWrappers wrappers, Test testInstance, int
...
+                var vtbl = Marshal.PtrToStructure<ITestVtbl>(inst.Vtbl);
+                var setValue = (delegate* unmanaged<IntPtr, int, int>)vtbl.SetValue;
+
+                Assert.Equal(0, setValue(itestPtr, 42));
+                Assert.Equal(42, Test.Resurrected.GetValue());
+
+                // release for QueryInterface
+                Assert.Equal(1, Marshal.Release(itestPtr));
+                // release for GetOrCreateComInterfaceForObject
+                Assert.Equal(0, Marshal.Release(itestPtr));
```
[the remark this answers is not in the document]
Comment (AustinWise) [reply src/tests/Interop/COM/ComWrappers/API/Program.cs]: Thanks, I fixed it.

**Comment by AustinWise** [comment]
I updated how the responsibility is split for keeping objects alive and collecting them. I think this new design is much easier to understand. The description of the PR has been updated.

The main problem this PR does not address is the issue of extra ManagedObjectWrappers hanging around for the lifetime of a wrapped object when there is a race condition adding values to the `_ccwTable`.

**Comment by AustinWise** [comment]
I realized that I did not need to use an extra handle to monitor the lifetime of the holder object from the releaser object. Instead I could just use the ref counted handle. That saves allocating an extra GC handle.

**Review by jkoritzinsky — APPROVED** [review]
This looks good to me, but I want @AaronRobinsonMSFT to take another look.

**Comment by AaronRobinsonMSFT** [comment]
> The key problem is the conditional weak table will keep the ManagedObjectWrapperHolder alive but not prevent it from being finalized. 

My biggest issue here is the above statement. If an object is "alive", I assume "reachable", then it will not be finalized. The GC only adds objects to the finalizer queue if they are no longer reachable so I am confused by this statement.

**Comment by AaronRobinsonMSFT** [comment]
> > The key problem is the conditional weak table will keep the ManagedObjectWrapperHolder alive but not prevent it from being finalized.
> 
> My biggest issue here is the above statement. If an object is "alive", I assume "reachable", then it will not be finalized. The GC only adds objects to the finalizer queue if they are no longer reachable so I am confused by this statement.

Oh right. The conditional weak table doesn't keep it alive, merely a reference to acquire. Its state is indeterminant in this case and that is the issue you are attempting to mitigate. This is simply loose terminology.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**